### PR TITLE
color grouping: default group by experiment on comparison view

### PR DESCRIPTION
**Pull request by japie1235813**
* Motivation for features / changes
When it's experiment comparison view, we want to color group by experiment even thought the user has not explicitly set it.

* Technical description of changes
   - check and update `initialGroupBy` on `createRouteContextedState`. It checks the routekind everytime the route changes.
   - Also fixes groupruns bug: when userSetGroupByKey is 0, it means group by run. However the previous logic does not correctly check null and misjudge group by run as invalid.

### CODE REVIEW

File: tensorboard/webapp/app_routing/testing.ts [hunk 1]
```diff
@@ -36,6 +36,22 @@ export function buildRoute(routeOverride: Partial<Route> = {}): Route {
   };
 }
 
+export function buildCompareRoute(
+  experimentIds: string[],
```
Comment (stephanwlee) [inline tensorboard/webapp/app_routing/testing.ts:40]: `aliasAndExperimentIds` would be more accurate name.

File: tensorboard/webapp/runs/store/runs_reducers.ts [hunk 2]
```diff
@@ -224,13 +235,19 @@ const dataReducer: ActionReducer<RunsDataState, Action> = createReducer(
     const groupKeyToColorString = new Map(state.groupKeyToColorString);
     const defaultRunColorForGroupBy = new Map(state.defaultRunColorForGroupBy);
 
+    let groupby = state.initialGroupBy;
```
Comment (stephanwlee) [inline tensorboard/webapp/runs/store/runs_reducers.ts:238]: groupby -> groupBy

File: tensorboard/webapp/runs/store/runs_reducers.ts [hunk 3]
```diff
@@ -224,13 +235,19 @@ const dataReducer: ActionReducer<RunsDataState, Action> = createReducer(
...
+      groupby = createGroupBy(
+        state.userSetGroupByKey,
+        state.colorGroupRegexString
+      );
+    }
     const groups = groupRuns(
-      state.userSetGroupByKey
-        ? createGroupBy(state.userSetGroupByKey, state.colorGroupRegexString)
-        : state.initialGroupBy,
+      groupby,
```
Comment (stephanwlee) [inline tensorboard/webapp/runs/store/runs_reducers.ts:246]: Just wondering, do you need this change? The code looks about the same to me.

File: tensorboard/webapp/runs/store/runs_reducers_test.ts [hunk 4]
```diff
@@ -965,5 +966,67 @@ describe('runs_reducers', () => {
...
+        state,
+        navigated({
+          before: buildCompareRoute(['eid1:run1', 'eid2:run2']),
+          after: buildRoute({
+            routeKind: RouteKind.EXPERIMENT,
+          }),
+        })
+      );
+
+      expect(nextState.data.initialGroupBy.key).toEqual(GroupByKey.RUN);
```
Comment (stephanwlee) [inline tensorboard/webapp/runs/store/runs_reducers_test.ts:1029]: toEqual -> toBe. Here and above.

File: tensorboard/webapp/runs/store/runs_reducers_test.ts [hunk 5]
```diff
@@ -965,5 +966,67 @@ describe('runs_reducers', () => {
...
+      const state3 = runsReducers.reducers(
+        state2,
+        navigated({
+          before: buildCompareRoute(['eid3:run3', 'eid4:run4']),
+          after: buildCompareRoute(['eid1:run1', 'eid2:run2']),
+        })
+      );
+
+      expect(state3.data.initialGroupBy.key).toEqual(GroupByKey.EXPERIMENT);
+      expect(state3.data.userSetGroupByKey).toEqual(GroupByKey.RUN);
```
Comment (stephanwlee) [inline tensorboard/webapp/runs/store/runs_reducers_test.ts:1011]: Consider wrapping all new tests around a contextual `describe` block and say something like "when freshly navigating"

File: tensorboard/webapp/runs/store/runs_reducers_test.ts [hunk 6]
```diff
@@ -965,5 +966,67 @@ describe('runs_reducers', () => {
...
+            routeKind: RouteKind.EXPERIMENT,
+          }),
+          after: buildCompareRoute(['eid1:run1', 'eid1:run2']),
+        })
+      );
+
+      expect(nextState.data.initialGroupBy.key).toEqual(GroupByKey.EXPERIMENT);
+    });
+
+    it('preserves userSetGroupByKey on comparision view navigation', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/runs/store/runs_reducers_test.ts:988]: 'preserves userSetGroupByKey on comparision view navigation' -> 'preserves userSetGroupByKey set on prior navigation to the same route'

File: tensorboard/webapp/runs/store/runs_reducers.ts (hunk 3 above)
Comment (japie1235813) [reply tensorboard/webapp/runs/store/runs_reducers.ts:246]: what was wrong is,
When `userSetGroupByKey = GroupByKey.RUN` it's `0` which would result in `groups` to be `state.initialGroupBy`
I mean to check null but what I wrote did not.

File: tensorboard/webapp/app_routing/testing.ts (hunk 1 above)
Comment (stephanwlee) [reply tensorboard/webapp/app_routing/testing.ts:40]: This is not resolved.

**Review by stephanwlee — APPROVED** [review]

File: tensorboard/webapp/app_routing/testing.ts (hunk 1 above)
Comment (japie1235813) [reply tensorboard/webapp/app_routing/testing.ts]: oops thanks for catching this. forgot to git add.

**Pull request merged**